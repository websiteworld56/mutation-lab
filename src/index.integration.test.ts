const { graphql } = require('@octokit/graphql');
import { table } from 'table';
import { executeTask } from './index';
import { when } from 'jest-when';
const core = require('@actions/core');
const github = require('@actions/github');
const { Octokit } = require('@octokit/rest');

jest.mock('@actions/core');
jest.mock('@actions/github');
jest.mock('@octokit/graphql');
jest.mock('@octokit/rest');
jest.mock('table');

const restMock = {
    pulls: {
        createReplyForReviewComment: jest.fn(),
    },
};
(Octokit as any).mockImplementation(() => restMock);

jest.setTimeout(10000000);

describe('Integration Test', () => {
    beforeEach(() => {
        jest.clearAllMocks();

        // Mock Inputs
        when(core.getInput).calledWith('token').mockReturnValueOnce('some-token');
        when(core.getInput).calledWith('test-command').mockReturnValueOnce('npm test');
        when(core.getInput).calledWith('test-command-directory').mockReturnValueOnce('./demo-app');
        when(core.getInput).calledWith('test-command-timeout').mockReturnValueOnce('5000');
        when(core.getInput).calledWith('auto-resolve').mockReturnValueOnce(undefined);
        github['context'] = {
            payload: {
                pull_request: {number: 123},
                repository: {
                    owner: {login: 'some-owner'},
                    name: 'some-repo-name'
                }
            }
        };
    })

    it('should add two landmines and catch them all', async () => {
        graphql.mockReturnValueOnce({
            repository: {
                pullRequest: {
                    reviewThreads: {
                        nodes: [{
                            id: 'MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy',
                            isResolved: false,
                            line: 12,
                            originalLine: 12,
                            originalStartLine: null,
                            comments: {
                                nodes: [{
                                    id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                    body: "ðŸ’£\r\n```suggestion\r\n    return null;\r\n```",
                                    path: "demo-app/insertion-sort.js"
                                }]
                            }
                        },{
                            id: 'MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy',
                            isResolved: false,
                            line: 12,
                            originalLine: 12,
                            originalStartLine: null,
                            comments: {
                                nodes: [{
                                    id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                    body: "/bomb\r\n```suggestion\r\n    return null;\r\n```",
                                    path: "demo-app/insertion-sort.js"
                                },{
                                    id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                    body: "ðŸ’¥ Bomb not defused",
                                    path: "demo-app/insertion-sort.js"
                                }]
                            }
                        }]
                    }
                }
            }
        });

        await executeTask();

        expect(core.setFailed).not.toHaveBeenCalled();
        expect(restMock.pulls.createReplyForReviewComment).toMatchSnapshot();
        expect(graphql.mock.calls).toMatchSnapshot();
        expect((table as any).mock.calls).toMatchSnapshot('bomb report');
    });

    it('should add a landmine but not catch it', async () => {
        graphql.mockReturnValueOnce({
            repository: {
                pullRequest: {
                    reviewThreads: {
                        nodes: [{
                            id: 'MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy',
                            isResolved: false,
                            line: 3,
                            originalLine: 3,
                            originalStartLine: 2,
                            comments: {
                                nodes: [
                                    {
                                        id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                        body: "ðŸ’£\r\n```suggestion\r\n    let length = inputArr.length; console.log('here');\r\n    for (let i = 1; i < length; i++) {\r\n```",
                                        path: "demo-app/insertion-sort.js"
                                    }
                                ]
                            }
                        }]
                    }
                }
            }
        });

        await executeTask();

        expect(core.setFailed).toHaveBeenCalledWith('There was at least one bomb that was not defused.');
        expect(restMock.pulls.createReplyForReviewComment).toMatchSnapshot();
        expect(graphql.mock.calls).toMatchSnapshot();
        expect((table as any).mock.calls).toMatchSnapshot('bomb report');
    });

    it('should add a landmine but not catch it because the execution timed out', async () => {
        graphql.mockReturnValueOnce({
            repository: {
                pullRequest: {
                    reviewThreads: {
                        nodes: [{
                            id: 'MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy',
                            isResolved: false,
                            line: 8,
                            originalLine: 8,
                            originalStartLine: 7,
                            comments: {
                                nodes: [
                                    {
                                        id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                        body: "ðŸ’£\r\n```suggestion\r\n            console.log('infinite loop!');\r\n```",
                                        path: "demo-app/insertion-sort.js"
                                    }
                                ]
                            }
                        }]
                    }
                }
            }
        });

        await executeTask();

        expect(core.setFailed).toHaveBeenCalledWith('There was at least one bomb that was not defused.');
        expect(restMock.pulls.createReplyForReviewComment).toMatchSnapshot();
        expect(graphql.mock.calls).toMatchSnapshot();
        expect((table as any).mock.calls).toMatchSnapshot('bomb report');
    });

    it('should not find any landmines because there are no threads', async () => {
        graphql.mockReturnValueOnce({
            repository: {
                pullRequest: {
                    reviewThreads: {
                        nodes: []
                    }
                }
            }
        });

        await executeTask();

        expect(core.setFailed).not.toHaveBeenCalled();
        expect(restMock.pulls.createReplyForReviewComment).toMatchSnapshot();
        expect(graphql.mock.calls).toMatchSnapshot();
        expect(table).not.toHaveBeenCalled();
    });

    it('should not find any landmines because there are no active threads', async () => {
        graphql.mockReturnValueOnce({
            repository: {
                pullRequest: {
                    reviewThreads: {
                        nodes: [{
                            id: 'MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy',
                            isResolved: true,
                            line: 12,
                            originalLine: 12,
                            originalStartLine: null,
                            comments: {
                                nodes: [{
                                    id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                    body: "ðŸ’£\r\n```suggestion\r\n    return null;\r\n```",
                                    path: "demo-app/insertion-sort.js"
                                }]
                            }
                        },{
                            id: 'MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy',
                            isResolved: false,
                            line: 12,
                            originalLine: 12,
                            originalStartLine: null,
                            comments: {
                                nodes: [{
                                    id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                    body: "/bomb\r\n```suggestion\r\n    return null;\r\n```",
                                    path: "demo-app/insertion-sort.js"
                                },{
                                    id: "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ3NzQ4NDE5Ng==",
                                    body: "âœ… Bomb successfully defused.",
                                    path: "demo-app/insertion-sort.js"
                                }]
                            }
                        }]
                    }
                }
            }
        });

        await executeTask();

        expect(core.setFailed).not.toHaveBeenCalled();
        expect(restMock.pulls.createReplyForReviewComment).toMatchSnapshot();
        expect(graphql.mock.calls).toMatchSnapshot();
        expect(table).not.toHaveBeenCalled();
    });
});
