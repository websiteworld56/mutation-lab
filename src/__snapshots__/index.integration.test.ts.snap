// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Test should add a landmine but not catch it 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "body": "ðŸ’¥ Bomb not defused. Please adjust your test to catch the error",
        "comment_id": "477484196",
        "owner": "some-owner",
        "pull_number": 123,
        "repo": "some-repo-name",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it 2`] = `
Array [
  Array [
    "
        {
          repository(name: \\"some-repo-name\\", owner: \\"some-owner\\") {
            pullRequest(number: 123) {
              reviewThreads(last: 100) {
                nodes {
                  id
                  isResolved
                  line
                  originalLine
                  originalStartLine
                  comments(last: 100) {
                    nodes {
                      id
                      body
                      path
                    }
                  }
                }
              }
            }
          }
        }",
    Object {
      "headers": Object {
        "authorization": "token some-token",
      },
    },
  ],
]
`;

exports[`Integration Test should add a landmine but not catch it because the execution timed out 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "body": "ðŸ’¥ Bomb not defused. Please adjust your test to catch the error",
        "comment_id": "477484196",
        "owner": "some-owner",
        "pull_number": 123,
        "repo": "some-repo-name",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it because the execution timed out 2`] = `
Array [
  Array [
    "
        {
          repository(name: \\"some-repo-name\\", owner: \\"some-owner\\") {
            pullRequest(number: 123) {
              reviewThreads(last: 100) {
                nodes {
                  id
                  isResolved
                  line
                  originalLine
                  originalStartLine
                  comments(last: 100) {
                    nodes {
                      id
                      body
                      path
                    }
                  }
                }
              }
            }
          }
        }",
    Object {
      "headers": Object {
        "authorization": "token some-token",
      },
    },
  ],
]
`;

exports[`Integration Test should add a landmine but not catch it because the execution timed out: bomb report 1`] = `
Array [
  Array [
    Array [
      Array [
        "File Name",
        "Line Numbers",
        "Bomb Defused",
      ],
      Array [
        "insertion-sort.js",
        "7-8",
        "no",
      ],
    ],
  ],
]
`;

exports[`Integration Test should add a landmine but not catch it: bomb report 1`] = `
Array [
  Array [
    Array [
      Array [
        "File Name",
        "Line Numbers",
        "Bomb Defused",
      ],
      Array [
        "insertion-sort.js",
        "2-3",
        "no",
      ],
    ],
  ],
]
`;

exports[`Integration Test should add two landmines and catch them all 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "body": "âœ… Successfully defused bomb",
        "comment_id": "477484196",
        "owner": "some-owner",
        "pull_number": 123,
        "repo": "some-repo-name",
      },
    ],
    Array [
      Object {
        "body": "âœ… Successfully defused bomb",
        "comment_id": "477484196",
        "owner": "some-owner",
        "pull_number": 123,
        "repo": "some-repo-name",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add two landmines and catch them all 2`] = `
Array [
  Array [
    "
        {
          repository(name: \\"some-repo-name\\", owner: \\"some-owner\\") {
            pullRequest(number: 123) {
              reviewThreads(last: 100) {
                nodes {
                  id
                  isResolved
                  line
                  originalLine
                  originalStartLine
                  comments(last: 100) {
                    nodes {
                      id
                      body
                      path
                    }
                  }
                }
              }
            }
          }
        }",
    Object {
      "headers": Object {
        "authorization": "token some-token",
      },
    },
  ],
  Array [
    "
              mutation {
                  resolveReviewThread(input: {threadId: \\"MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy\\"}) {
                    clientMutationId
                  }
              }
            ",
    Object {
      "headers": Object {
        "authorization": "token some-token",
      },
    },
  ],
  Array [
    "
              mutation {
                  resolveReviewThread(input: {threadId: \\"MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjk4NTE0MDI4OnYy\\"}) {
                    clientMutationId
                  }
              }
            ",
    Object {
      "headers": Object {
        "authorization": "token some-token",
      },
    },
  ],
]
`;

exports[`Integration Test should add two landmines and catch them all: bomb report 1`] = `
Array [
  Array [
    Array [
      Array [
        "File Name",
        "Line Numbers",
        "Bomb Defused",
      ],
      Array [
        "insertion-sort.js",
        12,
        "yes",
      ],
      Array [
        "insertion-sort.js",
        12,
        "yes",
      ],
    ],
  ],
]
`;

exports[`Integration Test should not find any landmines because there are no active threads 1`] = `[MockFunction]`;

exports[`Integration Test should not find any landmines because there are no active threads 2`] = `
Array [
  Array [
    "
        {
          repository(name: \\"some-repo-name\\", owner: \\"some-owner\\") {
            pullRequest(number: 123) {
              reviewThreads(last: 100) {
                nodes {
                  id
                  isResolved
                  line
                  originalLine
                  originalStartLine
                  comments(last: 100) {
                    nodes {
                      id
                      body
                      path
                    }
                  }
                }
              }
            }
          }
        }",
    Object {
      "headers": Object {
        "authorization": "token some-token",
      },
    },
  ],
]
`;

exports[`Integration Test should not find any landmines because there are no threads 1`] = `[MockFunction]`;

exports[`Integration Test should not find any landmines because there are no threads 2`] = `
Array [
  Array [
    "
        {
          repository(name: \\"some-repo-name\\", owner: \\"some-owner\\") {
            pullRequest(number: 123) {
              reviewThreads(last: 100) {
                nodes {
                  id
                  isResolved
                  line
                  originalLine
                  originalStartLine
                  comments(last: 100) {
                    nodes {
                      id
                      body
                      path
                    }
                  }
                }
              }
            }
          }
        }",
    Object {
      "headers": Object {
        "authorization": "token some-token",
      },
    },
  ],
]
`;
